### PR TITLE
fix: update `Paginate` to use `FilterPaginate` in `ClientStates` and `ConnectionChannels` grpc queries

### PR DESCRIPTION
**Pull request by charleenfei**
## Description

closes: #2935 

### Commit Message / Changelog Entry

```bash
fix: update `Paginate` to use `FilterPaginate` in `ClientStates` and `ConnectionChannels` grpc queries
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/02-client/keeper/grpc_query.go [hunk 1]
```diff
@@ -63,25 +63,26 @@ func (q Keeper) ClientStates(c context.Context, req *types.QueryClientStatesRequ
 	clientStates := types.IdentifiedClientStates{}
 	store := prefix.NewStore(ctx.KVStore(q.storeKey), host.KeyClientStorePrefix)
 
-	pageRes, err := query.Paginate(store, req.Pagination, func(key, value []byte) error {
+	pageRes, err := query.FilteredPaginate(store, req.Pagination, func(key, value []byte, accumulate bool) (bool, error) {
```
Comment (crodriguezvega) [inline modules/core/02-client/keeper/grpc_query.go:66]: Would it be possible to modify [the existing test](https://github.com/cosmos/ibc-go/blob/main/modules/core/02-client/keeper/grpc_query_test.go#L131-L154) to check for the `total` value in the `pagination` field of the response (i.e. `total` doesn't account for `consensusStates`)?

Maybe could also add another test with two clients that sets `limit` to 1 and checks that the `nextKey` in `pagination` points to a key with the format `/<client-id>/clientState`. But I understand that maybe this is a bit too much, and anyway maybe it's testing functionality from `Paginate`, not really our gRPC endpoint, so we can also skip it.

File: modules/core/04-channel/keeper/grpc_query.go [hunk 2]
```diff
@@ -96,26 +96,27 @@ func (q Keeper) ConnectionChannels(c context.Context, req *types.QueryConnection
 	channels := []*types.IdentifiedChannel{}
 	store := prefix.NewStore(ctx.KVStore(q.storeKey), []byte(host.KeyChannelEndPrefix))
 
-	pageRes, err := query.Paginate(store, req.Pagination, func(key, value []byte) error {
+	pageRes, err := query.FilteredPaginate(store, req.Pagination, func(key, value []byte, accumulate bool) (bool, error) {
```
Comment (crodriguezvega) [inline modules/core/04-channel/keeper/grpc_query.go:99]: Same comment: maybe it's possible to modify one of the existing [tests](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/keeper/grpc_query_test.go#L203)?

**Review by crodriguezvega — COMMENTED** [review]
Thank you for fixing this, @charleenfei!

**Review by damiannolan — COMMENTED** [review]
Looks great, agree with @crodriguezvega, a couple of test cases before we merge would be awesome tho! 
Thanks for picking this up! :)

File: modules/core/02-client/keeper/grpc_query_test.go [hunk 3]
```diff
@@ -377,11 +387,11 @@ func (suite *KeeperTestSuite) TestQueryConsensusStates() {
 			if tc.expPass {
 				suite.Require().NoError(err)
 				suite.Require().NotNil(res)
+				suite.Require().Equal(tc.expectedConsensusStates, res.Pagination.Total)
 				suite.Require().Equal(len(expConsensusStates), len(res.ConsensusStates))
```
Comment (damiannolan) [inline modules/core/02-client/keeper/grpc_query_test.go:391]: Aren't these checks essentially the same?

Comment (charleenfei) [reply modules/core/02-client/keeper/grpc_query_test.go:391]: true actually, i guess this test was passing bc we were reconstructing the situation and did not have other items in the query resp, whereas in real life without `FilterPaginate` other items were being returned in addition.

Comment (charleenfei) [reply modules/core/02-client/keeper/grpc_query_test.go:391]: i can add some other items maybe to the update

File: modules/core/04-channel/keeper/grpc_query_test.go [hunk 4]
```diff
@@ -193,6 +197,7 @@ func (suite *KeeperTestSuite) TestQueryChannels() {
 			if tc.expPass {
 				suite.Require().NoError(err)
 				suite.Require().NotNil(res)
+				suite.Require().Equal(tc.expectedChannels, res.Pagination.Total)
 				suite.Require().Equal(expChannels, res.Channels)
 				suite.Require().Equal(len(expChannels), int(res.Pagination.Total))
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/grpc_query_test.go:202]: Sorry for slowing this down but I think this pretty much the same check as below too, right?
`suite.Require().Equal(len(expChannels), int(res.Pagination.Total))`

We can probably remove the `tc.expectedChannels`.

I think as long as the scenario from the original issue has been tested and covered then I'm happy to get this merged.

Comment (crodriguezvega) [reply modules/core/04-channel/keeper/grpc_query_test.go:202]: Yeah, I agree with @damiannolan and I think we can remove `tc.expectedChannels`.

File: modules/core/02-client/keeper/grpc_query_test.go [hunk 5]
```diff
@@ -328,6 +328,9 @@ func (suite *KeeperTestSuite) TestQueryConsensusStates() {
 				path := ibctesting.NewPath(suite.chainA, suite.chainB)
 				suite.coordinator.SetupClients(path)
 
+				path1 := ibctesting.NewPath(suite.chainA, suite.chainB)
```
Comment (crodriguezvega) [inline modules/core/02-client/keeper/grpc_query_test.go:331]: Do we need this extra path in the test?

File: modules/core/02-client/keeper/grpc_query_test.go [hunk 6]
```diff
@@ -163,10 +163,10 @@ func (suite *KeeperTestSuite) TestQueryClientStates() {
 			ctx := sdk.WrapSDKContext(suite.chainA.GetContext())
 
 			res, err := suite.chainA.QueryServer.ClientStates(ctx, req)
-
 			if tc.expPass {
 				suite.Require().NoError(err)
 				suite.Require().NotNil(res)
+				suite.Require().Equal(uint64(len(expClientStates)), res.Pagination.Total)
```
Comment (crodriguezvega) [inline modules/core/02-client/keeper/grpc_query_test.go:169]: Really a super nit, but my consistency freak would write the check the same way as in this [other test](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/keeper/grpc_query_test.go#L197). And maybe also put it as the last one (after `suite.Require().Equal(expClientStates.Sort(), res.ClientStates)`).

```suggestion
				suite.Require().Equal(len(expClientStates)), int(res.Pagination.Total))
```

**Review by crodriguezvega — APPROVED** [review]

**Pull request merged**